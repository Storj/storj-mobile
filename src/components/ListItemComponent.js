import {
    Text,
    View,
    StyleSheet,
    TouchableOpacity,
    Image,
    ScrollView,
    Platform,
    ProgressViewIOS,
    ProgressBarAndroid
} from 'react-native';
import React, {Component} from 'react';
import { connect } from 'react-redux';
import { getWidth, getHeight } from '../utils/adaptive';
import PropTypes from 'prop-types';

export default class ListItemComponent extends Component {
    constructor(props) {
        super(props);
        this.bucket = null;
        //console.log(props);
    }

    shouldComponentUpdate(nextProps) {
        if( this.props.item !== nextProps.item ||
            this.props.isSelectionModeEnabled !== nextProps.isSelectionModeEnabled ||
            this.props.isSelected !== nextProps.isSelected ||
            this.props.isItemActionsSelected !== nextProps.isItemActionsSelected ||
            this.props.isSingleItemSelected !== nextProps.isSingleItemSelected) 
                return true;
        
        //TODO: Optimize render performance
        return false;
    }

    deselectAllItems() {
        this.props.selectItemId(null);
        this.props.disableSelectionMode();
    }

    render() {
        let name = this.props.item ? this.props.item.getName() : "name";
        let isSelected = this.props.item ? this.props.item.isSelected: false;
        let props = this.props;

        return(
            <TouchableOpacity 
                style = { props.isItemActionsSelected ? [listItemStyles.listItemContainer, listItemStyles.itemSelected] : listItemStyles.listItemContainer }
                onPress = { () => {
                    if(props.isSingleItemSelected) {
                        this.deselectAllItems()
                    } else if(props.isSelectionModeEnabled) {
                        props.onSelectionPress(props.item); 
                    } else {
                        this.props.onPress({ bucketId: this.props.item.getId() });
                    }
                }}
                onLongPress = { () => { 
                    props.onLongPress(props.item);
                }}>
                    <View style = { listItemStyles.listItemContent }>
                        {
                            (() => {
                                if(props.isSelectionModeEnabled) {
                                    if(isSelected) {
                                        return(<Image style = { listItemStyles.selectedIcon } source = { require('../images/Icons/ListItemSelected.png') } />);    
                                    } else {
                                        return(<Image style = { listItemStyles.selectedIcon } source = { require('../images/Icons/ListItemUnselected.png') } />);
                                    }
                                }
                            })()
                        }
                        <View>
                            <Image style = { listItemStyles.itemTypeIcon } source = { props.listItemIcon } />
                        </View>
                        <View style = { listItemStyles.textWrapper }>
                            <Text numberOfLines = {1} style = { listItemStyles.mainTitleText }>{ props.item.getName() }</Text>
                            {
                                this.props.progress ? 
                                    Platform.select({
                                        ios: 
                                            <ProgressViewIOS 
                                                progress = { 0 }
                                                trackTintColor = { '#f2f2f2' }
                                                progressTintColor = { '#2794ff' } />,
                                        android:
                                            <ProgressBarAndroid    
                                                progress = { 0 } 
                                                styleAttr = { 'Horizontal' } 
                                                color = { '#2794FF' } 
                                                animating = {true} 
                                                indeterminate = { false } />
                                    }) : null
                            }
                        </View>
                        {
                            !props.isSelectionModeEnabled ? 
                                <TouchableOpacity 
                                    style = { listItemStyles.listItemActionsIconContainer } 
                                    onPress = { () => {                     
                                        if(props.isSingleItemSelected) {
                                            props.selectItemId(null);
                                            this.deselectAllItems()
                                        } else {
                                            props.onSingleItemSelected();
                                            props.selectItemId(props.item.getId());
                                            //props.onPress(props.item);
                                        }
                                    }}>
                                    <Image style = { listItemStyles.listItemActionsIcon } source = { require('../images/Icons/listItemActions.png') } />
                                </TouchableOpacity> : null
                        }
                    </View>
            </TouchableOpacity>
        );
    }
}   

//TODO: add comments for all this stuff
ListItemComponent.propTypes = {
    item: PropTypes.object,
    onLongPress: PropTypes.func,
    isSelectionModeEnabled: PropTypes.bool,
    isSingleItemSelected: PropTypes.bool,
    isItemActionsSelected: PropTypes.bool,
    selectItemId: PropTypes.func,
    onPress: PropTypes.func,
    progress: PropTypes.number,
    listItemIcon: PropTypes.number, //wtf?
    isFileLoading: PropTypes.bool,
    isSelected: PropTypes.bool,
    disableSelectionMode: PropTypes.func,
    onSingleItemSelected: PropTypes.func
};

const listItemStyles = StyleSheet.create({
    listItemContainer: {
        flex: 1,
        height: getHeight(55),
        paddingHorizontal: getWidth(10)
    },
    itemSelected: {
        backgroundColor: '#d4e6ff',
        borderRadius: getHeight(6)
    },
    listItemContent: {
        flex: 1,
        flexDirection: 'row',
        borderBottomWidth: 0.5,
        borderColor: 'rgba(56, 75, 101, 0.4)',
        alignItems: 'center',
        justifyContent: 'space-between'
    },
    textWrapper: {
        flex: 0.9
    },
    mainTitleText: {
        fontFamily: 'Montserrat-Regular',
        lineHeight: getHeight(20),
        fontSize: getHeight(16),
        color: '#384B65'
    },
    listItemActionsIconContainer: {        
        height: getHeight(55),
        width: getWidth(50),
        alignItems: 'flex-end',
        justifyContent: 'center'
    },
    itemTypeIcon: {
        width: getWidth(25),
        height: getHeight(22)
    },
    listItemActionsIcon: {
        width: getWidth(20),
        height: getWidth(20)
    },
    selectedIcon: {
        width: getWidth(20),
        height: getWidth(20)
    }
});