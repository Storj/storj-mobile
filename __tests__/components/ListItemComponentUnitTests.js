import {
    View,
    TouchableOpacity
} from 'react-native';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'
import ListItemComponent from '../../src/components/ListItemComponent';

Enzyme.configure({ adapter: new Adapter() });

describe('ListItemComponent', ()=> {

	it('renders correctly', () => {
		const wrapper = shallow(
			<ListItemComponent 
                item = {{ getName: () => {}, getStarred: () => {}, entity: { thumbnail: {} } }} />	
		);

		expect(wrapper).toMatchSnapshot();
	});

	it('triggers onPress correctly with SelectionModeEnabled and SingleItemSelected', () => {
        const selectItemIdSpy = jest.fn();
        const disableSelectionModeSpy = jest.fn();

		const wrapper = shallow(
            <ListItemComponent 
                item = {{ getName: () => {}, getStarred: () => {}, entity: { thumbnail: {} } }}
                isSelectionModeEnabled = { true }
                isSingleItemSelected = { true }
                selectItemId = { selectItemIdSpy }
                disableSelectionMode = { disableSelectionModeSpy } />
        );
        wrapper.find('TouchableOpacity').forEach(child => {
            child.simulate('Press');
        });

        expect(selectItemIdSpy.mock.calls.length).toBe(1);
        expect(disableSelectionModeSpy.mock.calls.length).toBe(1);
    }); 
    
    it('triggers onPress correctly with not SelectionModeEnabled and SingleItemSelected', () => {
        const selectItemIdSpy = jest.fn();
        const disableSelectionModeSpy = jest.fn();

		const wrapper = shallow(
			<ListItemComponent 
                item = {{ getName: () => {}, getStarred: () => {}, entity: { thumbnail: {} } }}
                isSelectionModeEnabled = { false }
                isSingleItemSelected = { true }
                selectItemId = { selectItemIdSpy }
                disableSelectionMode = { disableSelectionModeSpy } />
        );

        wrapper.find('TouchableOpacity').forEach(child => {
            child.simulate('Press');
        });

        expect(selectItemIdSpy.mock.calls.length).toBe(2);
		expect(disableSelectionModeSpy.mock.calls.length).toBe(2);
	}); 

    it('triggers onPress correctly with not SelectionModeEnabled and not SingleItemSelected', () => {
        const selectItemIdSpy = jest.fn();
        const onSingleItemSelectedSpy = jest.fn();
        const onSelectionPressSpy = jest.fn();
        const pressSpy = jest.fn();

		const wrapper = shallow(
			<ListItemComponent 
                item = {{ getName: () => {}, getStarred: () => {}, getId: () => {}, entity: {bucketId: null, thumbnail: {}} }}
                isSelectionModeEnabled = { false }
                isSingleItemSelected = { false }
                onSelectionPress = { onSelectionPressSpy }
                selectItemId = { selectItemIdSpy }
                onSingleItemSelected = { onSingleItemSelectedSpy }
                onPress = { pressSpy } />
        );

        wrapper.find('TouchableOpacity').forEach(child => {
            child.simulate('Press');
        });

        expect(selectItemIdSpy.mock.calls.length).toBe(1);
        expect(onSelectionPressSpy.mock.calls.length).toBe(1);
        expect(onSingleItemSelectedSpy.mock.calls.length).toBe(1);
		expect(pressSpy.mock.calls.length).toBe(1);
    });
    
    it('triggers onLongPress correctly ', () => {
        const longPressSpy = jest.fn();

		const wrapper = shallow(
            <ListItemComponent 
                item = {{ getName: () => {}, getStarred: () => {}, getId: () => {}, entity: {bucketId: null} }}
                onLongPress = { longPressSpy } />
        );

        wrapper.find('TouchableOpacity').forEach(child => {
            child.simulate('LongPress');
        });

		expect(longPressSpy.mock.calls.length).toBe(1);
	});
});

