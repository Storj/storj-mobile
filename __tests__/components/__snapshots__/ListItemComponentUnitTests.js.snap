// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemComponent renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ListItemComponent
    item={
      Object {
        "entity": Object {
          "thumbnail": Object {},
        },
        "getName": [Function],
        "getStarred": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activeOpacity": 0.2,
      "children": <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0.5,
            "borderColor": "rgba(56, 75, 101, 0.4)",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginHorizontal": 20,
          }
        }
      >
        <View>
          <Image
            source={
              Object {
                "uri": "data:image/png;base64,[object Object]",
              }
            }
            style={
              Object {
                "height": 44,
                "width": 50,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 0.9,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Object {
                "color": "#384B65",
                "fontFamily": "Montserrat-Regular",
                "fontSize": 32,
              }
            }
          />
        </View>
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
          style={
            Object {
              "alignItems": "flex-end",
              "height": 110,
              "justifyContent": "center",
              "width": 100,
            }
          }
        >
          <Image
            source="listItemActions.png"
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
          />
        </TouchableOpacity>
      </View>,
      "onLongPress": [Function],
      "onPress": [Function],
      "style": Object {
        "flex": 1,
        "height": 110,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          undefined,
          <View>
            <Image
              source={
                Object {
                  "uri": "data:image/png;base64,[object Object]",
                }
              }
              style={
                Object {
                  "height": 44,
                  "width": 50,
                }
              }
            />
          </View>,
          <View
            style={
              Object {
                "flex": 0.9,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "color": "#384B65",
                  "fontFamily": "Montserrat-Regular",
                  "fontSize": 32,
                }
              }
            />
          </View>,
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
            style={
              Object {
                "alignItems": "flex-end",
                "height": 110,
                "justifyContent": "center",
                "width": 100,
              }
            }
          >
            <Image
              source="listItemActions.png"
              style={
                Object {
                  "height": 40,
                  "width": 40,
                }
              }
            />
          </TouchableOpacity>,
        ],
        "style": Object {
          "alignItems": "center",
          "borderBottomWidth": 0.5,
          "borderColor": "rgba(56, 75, 101, 0.4)",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
        },
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Image
              source={
                Object {
                  "uri": "data:image/png;base64,[object Object]",
                }
              }
              style={
                Object {
                  "height": 44,
                  "width": 50,
                }
              }
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "source": Object {
                "uri": "data:image/png;base64,[object Object]",
              },
              "style": Object {
                "height": 44,
                "width": 50,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "color": "#384B65",
                    "fontFamily": "Montserrat-Regular",
                    "fontSize": 32,
                  }
                }
              />,
              null,
            ],
            "style": Object {
              "flex": 0.9,
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "accessible": true,
                "allowFontScaling": true,
                "children": undefined,
                "ellipsizeMode": "tail",
                "numberOfLines": 1,
                "style": Object {
                  "color": "#384B65",
                  "fontFamily": "Montserrat-Regular",
                  "fontSize": 32,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            null,
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "activeOpacity": 0.2,
            "children": <Image
              source="listItemActions.png"
              style={
                Object {
                  "height": 40,
                  "width": 40,
                }
              }
            />,
            "onPress": [Function],
            "style": Object {
              "alignItems": "flex-end",
              "height": 110,
              "justifyContent": "center",
              "width": 100,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "source": "listItemActions.png",
              "style": Object {
                "height": 40,
                "width": 40,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeOpacity": 0.2,
        "children": <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 0.5,
              "borderColor": "rgba(56, 75, 101, 0.4)",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginHorizontal": 20,
            }
          }
        >
          <View>
            <Image
              source={
                Object {
                  "uri": "data:image/png;base64,[object Object]",
                }
              }
              style={
                Object {
                  "height": 44,
                  "width": 50,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "flex": 0.9,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "color": "#384B65",
                  "fontFamily": "Montserrat-Regular",
                  "fontSize": 32,
                }
              }
            />
          </View>
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
            style={
              Object {
                "alignItems": "flex-end",
                "height": 110,
                "justifyContent": "center",
                "width": 100,
              }
            }
          >
            <Image
              source="listItemActions.png"
              style={
                Object {
                  "height": 40,
                  "width": 40,
                }
              }
            />
          </TouchableOpacity>
        </View>,
        "onLongPress": [Function],
        "onPress": [Function],
        "style": Object {
          "flex": 1,
          "height": 110,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            undefined,
            <View>
              <Image
                source={
                  Object {
                    "uri": "data:image/png;base64,[object Object]",
                  }
                }
                style={
                  Object {
                    "height": 44,
                    "width": 50,
                  }
                }
              />
            </View>,
            <View
              style={
                Object {
                  "flex": 0.9,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "color": "#384B65",
                    "fontFamily": "Montserrat-Regular",
                    "fontSize": 32,
                  }
                }
              />
            </View>,
            <TouchableOpacity
              activeOpacity={0.2}
              onPress={[Function]}
              style={
                Object {
                  "alignItems": "flex-end",
                  "height": 110,
                  "justifyContent": "center",
                  "width": 100,
                }
              }
            >
              <Image
                source="listItemActions.png"
                style={
                  Object {
                    "height": 40,
                    "width": 40,
                  }
                }
              />
            </TouchableOpacity>,
          ],
          "style": Object {
            "alignItems": "center",
            "borderBottomWidth": 0.5,
            "borderColor": "rgba(56, 75, 101, 0.4)",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginHorizontal": 20,
          },
        },
        "ref": null,
        "rendered": Array [
          undefined,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Image
                source={
                  Object {
                    "uri": "data:image/png;base64,[object Object]",
                  }
                }
                style={
                  Object {
                    "height": 44,
                    "width": 50,
                  }
                }
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "source": Object {
                  "uri": "data:image/png;base64,[object Object]",
                },
                "style": Object {
                  "height": 44,
                  "width": 50,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    Object {
                      "color": "#384B65",
                      "fontFamily": "Montserrat-Regular",
                      "fontSize": 32,
                    }
                  }
                />,
                null,
              ],
              "style": Object {
                "flex": 0.9,
              },
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "accessible": true,
                  "allowFontScaling": true,
                  "children": undefined,
                  "ellipsizeMode": "tail",
                  "numberOfLines": 1,
                  "style": Object {
                    "color": "#384B65",
                    "fontFamily": "Montserrat-Regular",
                    "fontSize": 32,
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              null,
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "activeOpacity": 0.2,
              "children": <Image
                source="listItemActions.png"
                style={
                  Object {
                    "height": 40,
                    "width": 40,
                  }
                }
              />,
              "onPress": [Function],
              "style": Object {
                "alignItems": "flex-end",
                "height": 110,
                "justifyContent": "center",
                "width": 100,
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "source": "listItemActions.png",
                "style": Object {
                  "height": 40,
                  "width": 40,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
